core/src/main/java/org/apache/druid/segment/column/TypeStrategies.java

```diff
@@ -0,0 +1,463 @@
+      final int length = buffer.getInt();
+      final byte[] blob = new byte[length];
+      buffer.get(blob, 0, length);
+      return StringUtils.fromUtf8(blob);
+    }
+
+    @Override
+    public int write(ByteBuffer buffer, String value, int maxSizeBytes)
+    {
+      final byte[] bytes = StringUtils.toUtf8(value);
+      final int max = Math.min(buffer.limit() - buffer.position(), maxSizeBytes);
+      final int sizeBytes = Integer.BYTES + bytes.length;
+      final int remaining = max - sizeBytes;
+      if (remaining >= 0) {
+        buffer.putInt(bytes.length);
+        buffer.put(bytes, 0, bytes.length);
+        return sizeBytes;
+      }
+      return remaining;
+    }
+
+    @Override
+    public int compare(String s, String s2)
+    {
+      // copy of lexicographical string comparator in druid processing
+      // Avoid comparisons for equal references
+      // Assuming we mostly compare different strings, checking s.equals(s2) will only make the comparison slower.
+      //noinspection StringEquality
+      if (s == s2) {
+        return 0;
+      }
+
+      return ORDERING.compare(s, s2);
+    }
+  }
+
+  public static final class ArrayTypeStrategy implements TypeStrategy<Object[]>
+  {
+    private final Comparator<Object> elementComparator;
+    private final TypeStrategy elementStrategy;
+
+    public ArrayTypeStrategy(TypeSignature<?> type)
+    {
+      this.elementStrategy = type.getElementType().getNullableStrategy();
+      this.elementComparator = Comparator.nullsFirst(elementStrategy);
+    }
+
+    @Override
+    public int estimateSizeBytes(Object[] value)
+    {
+      if (value == null) {
+        return 0;
+      }
+      return Integer.BYTES + Arrays.stream(value).mapToInt(elementStrategy::estimateSizeBytes).sum();
+    }
+
+    @Override
+    public Object[] read(ByteBuffer buffer)
+    {
+      final int arrayLength = buffer.getInt();
+      final Object[] array = new Object[arrayLength];
+      for (int i = 0; i < arrayLength; i++) {
+        array[i] = elementStrategy.read(buffer);
+      }
+      return array;
+    }
+
+    @Override
+    public int write(ByteBuffer buffer, Object[] value, int maxSizeBytes)
+    {
+      final int max = Math.min(buffer.limit() - buffer.position(), maxSizeBytes);
+      int sizeBytes = Integer.BYTES;
+      int remaining = max - sizeBytes;
+      if (remaining < 0) {
+        return remaining;
+      }
+      int extraNeeded = 0;
+
+      buffer.putInt(value.length);
+      for (Object o : value) {
+        int written = elementStrategy.write(buffer, o, remaining);
+        if (written < 0) {
+          extraNeeded += written;
+          remaining = 0;
+        } else {
+          sizeBytes += written;
+          remaining -= sizeBytes;
+        }
+      }
+      return extraNeeded < 0 ? extraNeeded : sizeBytes;
+    }
+
+    @Override
+    public int compare(Object[] o1, Object[] o2)
+    {
+      final int iter = Math.max(o1.length, o2.length);
```

Comment:
umm should this be math.min as might have a scenario we can get indexOutofBoundException ?

Reply:
oops - yes, good catch, will add some tests since there obviously aren't any 😅